### DISCUSSION TITLE
Return type hint for macro closures?

### DISCUSSION THREAD

**Discussion by rudiedirkx** [larastan/larastan#2120] [Q&A]
Can I `@return` hint a macro return type somehow? My macro closure has a php return type:

```php
Connection::macro('selectArray', function(string $query, array $bindings = []) : array {
```

but just `array` is not good enough. It's a complex array that I want to describe.

I tried a stub:

```php
namespace Illuminate\Database {
	class Connection {
		/**
		 * @return array{list<array<string, scalar>>, array<string, string>}
		 */
		public function selectArray(string $query, array $bindings = []) : array;
	}
}
```

but it seems to be unused. Or maybe Larastan overwrites it with its macro detection?

I call the macro from the facade:

```php
[$records, $columnTypes] = \DB::selectArray($query->full_sql, $params);
```

so I tried stubbing that too:

```php
namespace Illuminate\Support\Facades {
	class DB {
		/**
		 * @return array{list<array<string, scalar>>, array<string, string>}
		 */
		static public function selectArray(string $query, array $bindings = []) : array;
	}
}
```

but still no effect.

----

What's weird is that the stub doesn't even work for a non-macro method. If I comment the real macro, and call the stub in real code, I expected Phpstan to read it, but instead I get an error

> Call to an undefined static method Illuminate\Support\Facades\DB::selectArray().

Even though that's exactly the method I stubbed... Why doesn't Phpstan believe it exists?

----

PHP/Phpstan/Larastan can read the closure's phpdoc, so Larastan could use that. Probably even Phpstan could use that in `ClosureTypeFactory`, but it doesn't, so Larastan could do that in `MacroMethodsClassReflectionExtension`, so I can do:

```php
Connection::macro('selectArray',
	/**
	 * @return array{list<array<string, scalar>>, array<string, string>}
	 */
	function(string $query, array $bindings = []) : array {
```

**Comment by rudiedirkx** [comment]
Maybe I just don't understand stubs 😮 If I call the stubbed method from the object:

```php
$conn = \DB::connection();
[$records, $columnTypes] = $conn->selectArray($query->full_sql, $params);
```

Phpstan still doesn't know `selectArray()`:

> Call to an undefined method Illuminate\Database\Connection::selectArray().

but that's what I stubbed!:

```php
namespace Illuminate\Database {
	class Connection {
		/**
		 * @return array{list<array<string, scalar>>, array<string, string>}
		 */
		public function selectArray(string $query, array $bindings = []) : array;
	}
}
```

and I removed the `Connection::macro('selectArray', function(....` so it's not Larastan interfering. Why is my stub wrong? It is read, it work for other classes, in `Illuminate` too.

**Reply by calebdw** [reply]
I'm fairly certain that stubs are for overriding the PHPDocs of *existing methods*, not for adding methods to a class

**Reply by rudiedirkx** [reply]
Aaw =( I guess that's probably for the best 😆 But then how do I tell what `selectArray()` returns?

And `@method` should still work, right? That's for adding unknown methods. Or not allowed in stub files? Doesn't work in this case:

```php
namespace Illuminate\Database {
	/**
	 * @method array{list<array<string, scalar>>, array<string, string>} selectArray(string $query, array $bindings)
	 */
	class Connection {}
}
```

**Reply by rudiedirkx** [reply]
Yeah, adding `@method` on the real Laravel `Connection` class does work. But not in a stub file. M'kay. Bummer. Then Larastan has to be smarter.

**Reply by calebdw** [reply]
Have you actually registered your stub file with PHPStan?

**Reply by calebdw** [reply]
And it might not work if there's a conflicting Larastan stub for the same class (I don't think PHPStan supports multiple stubs for the same class)

**Reply by rudiedirkx** [reply]
Yes, I registered it. It works for other classes. It must be what you said, stubs only work for known methods, not for new. `@method` on the real Laravel class works, Phpstan stub does not. That's almost always a good thing, but not for Laravel macros, because they don't really exist. Phpstan understands the closure's phpdoc, so Larastan could use that, see OP end edit.

**Reply by rudiedirkx** [reply]
> And it might not work if there's a conflicting Larastan stub for the same class

I even removed the Larastan stubs, and it still didn't work, so it looks like the method has to already exist for Phpstan.

**Reply by calebdw** [reply]
I would expect an `@method` on a stub to work

**Reply by rudiedirkx** [reply]
How do you this? Or do you not use macros (with array io)?

**Reply by calebdw** [reply]
I've just suffered from poor types on macros---I never thought to try and use a stub, that's pretty neat

**Reply by rudiedirkx** [reply]
Real neat if it worked. But I like the inline phpdoc on the macro itself even better. Define the return type where the method is defined. Like how all methods work. TIL Closure can have phpdoc reflection, so Phpstan or Larastan could do this, but I definitely don't know how. Phpstan's `ClosureTypeFactory::fromClosureObject()` is crazy complicated.

**Comment by canvural** [comment]
When we are reading the macro definition we do not take into account any PHPDocs. Maybe it is doable. I don't know. But if it is [this](https://github.com/larastan/larastan/blob/86abf11dd531d3ad9aee21b1427a416b9db78503/src/Methods/MacroMethodsClassReflectionExtension.php#L134-L138) is the place to do it.

**Reply by rudiedirkx** [reply]
I don't understand why Phpstan doesn't do it. `ClosureTypeFactory::fromClosureObject()` could do it, but doesn't use a Closure's `@return`. I tried in `MacroMethodsClassReflectionExtension` but it was a mess, because I don't know Phpstan well enough (reading phpdoc with namespaces etc).

**Reply by canvural** [reply]
There is an [issue](https://github.com/phpstan/phpstan/issues/3770) for that. Also our case is a special one, it needs more manual work.
